externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/ExecuteStatement.scala

```diff
@@ -0,0 +1,172 @@
+
+  private var resultDescriptor: ResultDescriptor = _
+
+  private var columnInfos: List[ColumnInfo] = _
+
+  private var statementTimeoutCleaner: Option[ScheduledExecutorService] = None
+
+  override def getOperationLog: Option[OperationLog] = Option(operationLog)
+
+  @VisibleForTesting
+  override def setExecutor(executor: Executor): Unit = {
+    this.executor = executor
+  }
+
+  def setSessionId(sessionId: String): Unit = {
+    this.sessionId = sessionId
+  }
+
+  override protected def beforeRun(): Unit = {
+    OperationLog.setCurrentOperationLog(operationLog)
+    setState(OperationState.PENDING)
+    setHasResultSet(true)
+  }
+
+  override protected def afterRun(): Unit = {
+    OperationLog.removeCurrentOperationLog()
+  }
+
+  override protected def runInternal(): Unit = {
+    //    executeStatement()
+
+    addTimeoutMonitor()
+    if (shouldRunAsync) {
+      val asyncOperation = new Runnable {
+        override def run(): Unit = {
+          OperationLog.setCurrentOperationLog(operationLog)
+        }
+      }
+
+      try {
+        executeStatement()
+        val flinkSQLSessionManager = session.sessionManager
+        val backgroundHandle = flinkSQLSessionManager.submitBackgroundOperation(asyncOperation)
+        setBackgroundHandle(backgroundHandle)
+      } catch {
+        case rejected: RejectedExecutionException =>
+          setState(OperationState.ERROR)
+          val ke =
+            KyuubiSQLException("Error submitting query in background, query rejected", rejected)
+          setOperationException(ke)
+          throw ke
+      }
+    } else {
+      executeStatement()
+    }
+  }
+
+  private def executeStatement(): Unit = {
+    try {
+      setState(OperationState.RUNNING)
+      val operation: Operation = executor.parseStatement(sessionId, statement)
+      resultDescriptor = executor.executeQuery(sessionId, operation.asInstanceOf[QueryOperation])
+
+      val resultSchemaColumns: util.List[Column] =
+        resultDescriptor.getResultSchema.getColumns
+      columnInfos = new util.ArrayList[ColumnInfo]
+      for (column <- resultSchemaColumns.asScala) {
+        columnInfos.add(ColumnInfo.create(column.getName, column.getDataType.getLogicalType))
+      }
+
+      val resultID = resultDescriptor.getResultId
+
+      var rows: util.List[Row] = new util.ArrayList[Row]()
+
+      var loop = true
+      while (loop) {
+        Thread.sleep(50) // slow the processing down
+
+        val result = executor.snapshotResult(sessionId, resultID, 2)
+        if (result.getType eq TypedResult.ResultType.PAYLOAD) {
+          rows.clear()
+          IntStream.rangeClosed(1, result.getPayload).forEach((page: Int) => {
+            def foo(page: Int) = {
+              for (row <- executor.retrieveResultPage(resultID, page)) {
+                rows.add(row)
+              }
+            }
+
+            foo(page)
+          })
+        } else if (result.getType eq TypedResult.ResultType.EOS) loop = false
```

Comment:
Ideally, we should return a iterator and lazy fetch the next snapshot (or periodically update the snapshot), but it's ok for now, as we just tend to support the simplest query `select 1` in this initial PR.